src/cache_filesystem.cpp

```diff
@@ -432,17 +499,33 @@ int64_t CacheFileSystem::ReadImpl(FileHandle &handle, void *buffer, int64_t nr_b
 		return 0;
 	}
 
+	auto state = instance_state.lock();
+
 	// If the source filepath matches exclusion regex, skip caching.
 	const int64_t bytes_to_read = MinValue<int64_t>(nr_bytes, file_size - location);
-	if (CacheExclusionManager::GetInstance().MatchAnyExclusion(handle.GetPath())) {
+	auto *exclusion_mgr = GetExclusionManager(state.get());
+	if (exclusion_mgr && exclusion_mgr->MatchAnyExclusion(handle.GetPath())) {
 		auto &cache_handle = handle.Cast<CacheFileSystemHandle>();
 		internal_filesystem->Read(*cache_handle.internal_file_handle, buffer, nr_bytes, location);
 		return bytes_to_read;
 	}
 
 	// Leverage cache read manager.
-	cache_reader_manager.GetCacheReader()->ReadAndCache(handle, static_cast<char *>(buffer), location, bytes_to_read,
-	                                                    file_size);
+	auto *mgr = GetCacheReaderManager(state.get());
+	if (mgr) {
+		auto *reader = mgr->GetCacheReader();
+		if (reader) {
+			reader->ReadAndCache(handle, static_cast<char *>(buffer), location, bytes_to_read, file_size);
+		} else {
```

Comment:
I would prefer to keep the current implementation, which sets a `CacheReader` anyway, with `NoopCacheReader` as the fallback option, so we don't need the if-else branch here. WDYT?